src/libstd/sys/windows/stdio.rs

```diff
@@ -1,140 +1,226 @@
+        let read = self.read_u16s_fixup_surrogates(handle, &mut utf16_buf, amount)?;
+        let utf16 = &utf16_buf[..read];
+
+        // FIXME: it would be nice if we could directly decode into the buffer instead of doing an
+        //        allocation.
+        let data = match String::from_utf16(&utf16) {
+            Ok(utf8) => utf8.into_bytes(),
+            Err(..) => {
+                // We can't really do any better than forget all data and return an error.
+                return Err(io::Error::new(io::ErrorKind::InvalidData,
+                    "Windows stdin in console mode does not support non-UTF-16 input; \
+                    encountered unpaired surrogate"))
+            },
         };
-        let mut utf8 = self.utf8.lock().unwrap();
-        // Read more if the buffer is empty
-        if utf8.position() as usize == utf8.get_ref().len() {
-            let mut utf16 = vec![0u16; 0x1000];
-            let mut num = 0;
-            let mut input_control = readconsole_input_control(CTRL_Z_MASK);
-            cvt(unsafe {
-                c::ReadConsoleW(handle,
-                                utf16.as_mut_ptr() as c::LPVOID,
-                                utf16.len() as u32,
-                                &mut num,
-                                &mut input_control as c::PCONSOLE_READCONSOLE_CONTROL)
-            })?;
-            utf16.truncate(num as usize);
-            // FIXME: what to do about this data that has already been read?
-            let mut data = match String::from_utf16(&utf16) {
-                Ok(utf8) => utf8.into_bytes(),
-                Err(..) => return Err(invalid_encoding()),
-            };
-            if let Some(&last_byte) = data.last() {
-                if last_byte == CTRL_Z {
-                    data.pop();
-                }
+        buf.copy_from_slice(&data);
+        Ok(data.len())
+    }
+
+    // We assume that if the last `u16` is an unpaired surrogate they got sliced apart by our
+    // buffer size, and keep it around for the next read hoping to put them together.
+    // This is a best effort, and may not work if we are not the only reader on Stdin.
+    pub fn read_u16s_fixup_surrogates(&self, handle: c::HANDLE, buf: &mut [u16], mut amount: usize)
+        -> io::Result<usize>
+    {
+        // Insert possibly remaining unpaired surrogate from last read.
+        let mut start = 0;
+        if self.high_surrogate.get() != 0 {
+            buf[0] = self.high_surrogate.replace(0);
+            start = 1;
+            if amount == 1 {
+                // Special case: `Stdin::read` guarantees we can always read at least one new `u16`
+                // and combine it with an unpaired surrogate, because the UTF-8 buffer is at least
+                // 4 bytes.
+                amount = 2;
             }
-            *utf8 = Cursor::new(data);
         }
+        let mut amount = read_u16s(handle, &mut buf[start..amount])? + start;
```

Comment:
Yes, that was the intention. If the half surrogate was cause by the previous buffer being to small, I would expect the next read to succeed. If it errors, then the next successful read probably has nothing to do with this half surrogate.

All edge cases and best efforts though. Combined with a buffer of 8KiB, and Console not displaying surrogate pairs nicely, it will be quite special if this ever happens :smile:.